### PR TITLE
[WIP] Use V2 API of NiceHash

### LINKED ISSUE #47: `attack_hourly_cost` value no longer returned.
There is no longer a number value returned for `attack_hourly_cost`, each coin has a `'?'` string instead.

This is breaking howmanyconfs.com (https://github.com/lukechilds/howmanyconfs.com/issues/12) which relies on that value in it's own calculations.

Is it possible to add support for this metric back in?

Thanks!

### PR DESCRIPTION
**Pull request by szmarczak**
Fixes #47

### CODE REVIEW

File: crypto51/apis/nicehash.py [hunk 1]
```diff
@@ -16,17 +16,17 @@ class NiceHash:
...
+        self._global_stats = requests.get('https://api2.nicehash.com/main/api/v2/public/stats/global/current').json()
         self._algo_ids = self._get_algo_ids()
 
     def _get_algo_ids(self):
-        algorithms = requests.get('https://api.nicehash.com/api?method=multialgo.info').json()['result']['multialgo']
+        algorithms = requests.get('https://api2.nicehash.com/main/api/v2/public/simplemultialgo/info/').json()['miningAlgorithms']
         algo_ids = {}
-        for a in algorithms:
-            name = a['name'].lower()
+        for index, a in enumerate(algorithms):
```
Comment (szmarczak) [inline crypto51/apis/nicehash.py]: I'm not sure if this is the right way to get the index...

**Comment by szmarczak** [comment]
I think I must've done something wrong:

image

File: crypto51/apis/nicehash.py [hunk 2]
```diff
@@ -75,14 +78,18 @@ def _get_in_nicehash_units(self, algorithm, value):
...
     def get_orders(self, algorithm):
         index = self._get_algorithm_index(algorithm)
         if index is None:
             return []
-        resp = requests.get('https://api.nicehash.com/api?method=orders.get&location=1&algo={}'.format(index))
+        """
+        I have no idea what does location=1 mean:
+        https://api.nicehash.com/api?method=orders.get&location=1&algo={}
+        """
+        resp = requests.get('https://api2.nicehash.com/main/api/v2/public/orders/?alghoritm={}'.format(index))
```
Comment (lukechilds) [inline crypto51/apis/nicehash.py]: Typo here:
```suggestion
        resp = requests.get('https://api2.nicehash.com/main/api/v2/public/orders/?algorithm={}'.format(index))
```

**Comment by szmarczak** [comment]
image

not sure if this is the desired result

**Comment by lukechilds** [comment]
@szmarczak not quite, but at least it looks like the API's responding now.

Here's an (outdated) example of how it should look:

Screenshot 2019-10-10 at 18 32 33

https://web.archive.org/web/20190828223227/https://www.crypto51.app/

**NiceHash-able** is the amount of hashrate available on NiceHash as a % of the coins hashrate. e.g if the coin's hashrate is 50MH/s and there is 100MH/s available on NiceHash for that coins algorithm, the **NiceHash-able** value should be 50%. For Bitcoin this should be 0% because it's hashrate should be way higher than what NiceHash has available.

**1h Attack Cost** is the theoretical cost in $ to match the coins hashrate (regardless if enough hashrate actually exists on NiceHash) for 1 hour. e.g if a coins's hashrate is 50MH/s and NiceHash sells hashrate for that coin's algorithm for $1/MH/Day then the **1h Attack Cost** value should be $2 ($1*50MH/24H).

**Comment by szmarczak** [comment]
NiceHash has got the worst documentation I've ever seen.

**Comment by lukechilds** [comment]
Try applying this patch if you wanna speed up testing:

```diff
diff --git a/crypto51/apis/wtm.py b/crypto51/apis/wtm.py
index 5baff65..09fc89a 100644
--- a/crypto51/apis/wtm.py
+++ b/crypto51/apis/wtm.py
@@ -14,6 +14,10 @@ class WTM:
 
         # Filter out 'lagging'
         for coin in wtm["coins"]:
+            # Skip coins for dev
+            if wtm["coins"][coin]['tag'] not in ['BTC', 'ETH', 'BCH', 'BSV']:
+                print("Skip coin " + wtm["coins"][coin]['tag'])
+                continue
             print("Grabbing more data for "+coin)
             if wtm["coins"][coin]['lagging'] == False and wtm["coins"][coin]['status'] == "Active":
                 print("-> Coin isn't lagging and is active")
```

**Comment by lukechilds** [comment]
Definitely getting closer:

```json
{
  "coins": [
    {
      "hash_rate_pretty": "98,884 PH/s",
      "nicehash_market_link": "https://www.nicehash.com/marketplace/sha256",
      "attack_hourly_cost_pretty": "$674,558",
      "market_cap": 153214913546,
      "name": "Bitcoin",
      "algorithm": "SHA-256",
      "network_vs_rentable_ratio": 0.00045055691502976064,
      "rentable_price_btc": 0.019217787110140012,
      "symbol": "BTC",
      "rentable_price_units": "PH",
      "attack_hourly_cost": 674558.3475464842,
      "rentable_capacity": 44552764956608430,
      "rentable_price_usd_hour": "$6.82",
      "rentable_capacity_pretty": "45 PH/s",
      "block_time": 565,
      "minethecoin_link": "https://google.com",
      "hash_rate": 98883766890284190000,
      "market_cap_pretty": "$153.21 B"
    },
    {
      "hash_rate_pretty": "182 TH/s",
      "nicehash_market_link": "https://www.nicehash.com/marketplace/ethash",
      "attack_hourly_cost_pretty": "$102,467,738",
      "market_cap": 20764296384,
      "name": "Ethereum",
      "algorithm": "Ethash",
      "network_vs_rentable_ratio": 0.026970952813320955,
      "rentable_price_btc": 1588.8005597521988,
      "symbol": "ETH",
      "rentable_price_units": "TH",
      "attack_hourly_cost": 102467738.18451403,
      "rentable_capacity": 4900302770499.586,
      "rentable_price_usd_hour": "$563,975.87",
      "rentable_capacity_pretty": "5 TH/s",
      "block_time": 13.5643,
      "minethecoin_link": "https://google.com",
      "hash_rate": 181688159273310,
      "market_cap_pretty": "$20.76 B"
    },
    {
      "hash_rate_pretty": "2,407 PH/s",
      "nicehash_market_link": "https://www.nicehash.com/marketplace/sha256",
      "attack_hourly_cost_pretty": "$16,418",
      "market_cap": 4177903454,
      "name": "BitcoinCashABC",
      "algorithm": "SHA-256",
      "network_vs_rentable_ratio": 0.018512126530514098,
      "rentable_price_btc": 0.019217787110140012,
      "symbol": "BCH",
      "rentable_price_units": "PH",
      "attack_hourly_cost": 16417.720977497916,
      "rentable_capacity": 44552764956608430,
      "rentable_price_usd_hour": "$6.82",
      "rentable_capacity_pretty": "45 PH/s",
      "block_time": 604,
      "minethecoin_link": "https://google.com",
      "hash_rate": 2406680015025327600,
      "market_cap_pretty": "$4.18 B"
    }
}
```

BTC and BCH look about right but ETH is waaaay over. It’s saying it’ll cost $674,558 to attack BTC for an hour but $102,467,738 to attack ETH for an hour.

Are you sure you’re getting the Ethash price for ETH? I think the algo names come from the whattomine api and are then sent to the NiceHash API and they don’t map 1:1. Some of them were manually mapped over to the different API names, maybe the API name changed in the v2 api?

**Comment by szmarczak** [comment]
Umm, here's what it does:
1. 1. Fetch https://api2.nicehash.com/main/api/v2/mining/algorithms/
1. 2. Figure out what algorithm ID BTC uses: `35`
2. 1. Fetch https://api2.nicehash.com/main/api/v2/public/stats/global/current
2. 2. Figure out the price (`p`) for `a: 35`: `1.9009460835892252E-9`
2. 3. Don't know why, but need to multiply by 10^7: `0.019009460835892254`

And while it works for BTC, it doesn't work for ETH:  https://www.nicehash.com/pricing

**Comment by tdickman** [comment]
> Don't know why, but need to multiply by 10^7: 0.019009460835892254

I'm struggling to get the numbers to line up correctly, but the solution lies in this factor that I believe changed in v2 to vary based on the algorithm. Nicehash's documentation is pretty low quality, so I'm unclear what units the value from step 4 is in. It hints at the "marketFactor", but that doesn't appear to work.

**Comment by lukechilds** [comment]
Ok, I took a stab at this and I think I've figured out the NiceHash values.

Atleast I've managed to recalculate the values here https://www.nicehash.com/algorithm from the API with a little Node.js script:

Screenshot 2019-10-11 at 14 26 34

```
nicehash %
››› node index.js

================
Scrypt
================
{
  hashrateRaw: 9096767822535.084,
  hashrateReadable: '9.0698 TH',
  priceRaw: 0.000018345834970544845,
  priceReadable: '0.1835 BTC/TH/day'
}

================
SHA256
================
{
  hashrateRaw: 295532115613255500,
  hashrateReadable: '295.5321 PH',
  priceRaw: 1.9101996782399107e-9,
  priceReadable: '0.0191 BTC/PH/day'
}

================
X11
================
{
  hashrateRaw: 116606319159613.03,
  hashrateReadable: '116.6063 TH',
  priceRaw: 3.5674289395164286e-7,
  priceReadable: '0.0036 BTC/TH/day'
}
```

Let me check with more coins and see if I can calculate all the attack costs.

**Comment by lukechilds** [comment]
@tdickman I've implemented this in JavaScript along with some other bits at https://github.com/lukechilds/howmanyconfs.com.

I've also exposed the data as an API endpoint which you're free to use for your site if you wish. You should be able to make one request to https://howmanyconfs.com/api/data and get all the data you need for crypto51.app.

The data is formatted like this:

```json
[
	{
	  "symbol": "BTC",
	  "name": "Bitcoin",
	  "marketCap": 162901616855.04,
	  "price": 9032.17,
	  "priceCurrency": "USD",
	  "algorithm": "SHA-256",
	  "hashrate": 85764531598724940000,
	  "blockTimeInSeconds": 637,
	  "blockReward": 12.8048,
	  "hashrateCostPerSecond": 0.020200062912737998,
	  "nicehasheable": 0.05951142873236588
	}
	...
]
```

So you can do:

```js
const oneHourAttackCost = coin.hashrateCostPerSecond * HOUR_IN_SECONDS;
const nicehasheablePercentage = coin.nicehasheable * 100;
```

**Comment by tdickman** [comment]
I finally got this to work. For some reason dividing the value from https://api2.nicehash.com/main/api/v2/public/stats/global/current by (10^8) gives the value in BTC / H / day. I am not sure *why* this works, but I've verified against nicehash's [algorithm pages](https://www.nicehash.com/algorithm/sha256).

**Comment by szmarczak** [comment]
@tdickman Thank you for fixing this PR! :D I couldn't figure this out by myself.

**Comment by tdickman** [comment]
@szmarczak Thank you for all of your help, you did most of the work, I just pushed it across the finish line. Sorry it took me so long to get around to looking into it (> 2 months).

Now I really wish I knew where this darn constant came from.

**Pull request merged**